src/gallium/drivers/lima/lima_texture.c

```diff
@@ -91,7 +91,7 @@ lima_update_tex_desc(struct lima_context *ctx, struct lima_sampler_state *sample
 
    /* TODO: - do we need to align width/height to 16?
             - does hardware support stride different from width? */
-   width = prsc->width0;
+   width = align(prsc->width0, 16);
```

Comment:
That's probably not a right place to do it. We should align resource stride when it's allocated.

Reply:
I guess the resource stride is already aligned when allocation, but the descriptor setup need a way to tell the width and stride difference. Texture UV space [0, 1] is mapped to pixel space [0, width], but texture is stored in memory in stride X height way. So we have to program in the descriptor both width and stride value. I can't see where current code reflect this, so we need descriptor changes for the width != stride case.

Reply:
@yuq I don't think that hardware supports that.

Reply:
robclark said that it's pretty common limitation of hardware, so we have to workaround it in the driver.

Reply:
Again, let's dump to see how mali handle this case.

Reply:
I did the dump and mali actually uses stride == width. So 500x500 texture works fine if stride is 500 * bpp. However we're aligning stride in lima_resource_create() - since texture is created with PIPE_BIND_RENDER_TARGET set in templat->bind.

If I comment out alignment 500x500 texture works fine, but I guess we'll break render target in this case...

Reply:
Then another dump is needed to see the render target texture case: create a texture and attach to FBO for rendering, then use it as texture to render to the window surface.

Reply:
OK, if it does extra copy, it'll be way much slower.

Maybe I can modify the gbm-surface-fbo for X11 and go for a dump next week after I get back from vacation.

Reply:
EGL_KHR_platform_gbm is supported by recent malion e.g. rockchip, see https://github.com/xbmc/xbmc/pull/13746/files#diff-2b5179a78cb072ff496b13892efaa8b2R41 for a bit more info.

Reply:
@koenkooi I still can't get it working. I replaced EGL_PLATFORM_GBM_MESA with EGL_PLATFORM_GBM_KHR and now it fails in eglCreatePlatformWindowSurfaceEXT() - it returns EGL_NO_SURFACE.